notes/conds.md

```diff
@@ -0,0 +1,84 @@
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `some`       |`[Expr<Any[]>, Expr<Bool>] -> Bool`| Collection predicate                        | `["some", ["/to"], ["==", [], "hi@web.mail"]]`|
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+    ["var", ["?x", "?y"], // The `?` here is just by convention
+      ["args", "foo..bar.[].baz", "?x"],
+      [["some", "match"], "?x", "*@example.com"],
+      
+      ["args", "foo.quux", "?y"],
+      [">", "?y", 42],
+      ["<", "?x", "?y"]
+    ]
+  ]
+}
+```
+
+Of note:
+
+* Selector syntax based on JSONPath / JSON Poiniter / jQ
+  * `"foo..bar.[].baz"` ~ `foo.{*recursive descent*}.bar.{ALL}.baz`
+* `var` is like `fresh` in minikanren: introduces logic variables
```

Comment:
I would personally defer new scope creation until we really need it and keep things simpler in the meantime.

Reply:
Yeah I don't love it either. How about something like this:

```js
{
  iss: "...",
  var: ["?x", "?y"],
  policy: [...]
}
```

This forces one set of variables, which I think I would _always_ prefer

Reply:
Yeah I'm just going to make this change